messaging/src/main/java/org/axonframework/messaging/deadletter/DeadLetterQueue.java

```diff
@@ -0,0 +1,229 @@
+     * given {@link QueueIdentifier} is full.
+     *
+     * @param queueIdentifier The identifier of the queue to validate for.
+     * @return {@code true} either when {@link #maxQueues()} is reached or when the {@link #maxQueueSize()} is reached.
+     * Returns {@code false} otherwise.
+     */
+    boolean isFull(@Nonnull QueueIdentifier queueIdentifier);
+
+    /**
+     * The maximum number of distinct queues this dead-letter queue can hold. This comes down to the maximum number of
+     * unique {@link QueueIdentifier QueueIdentifiers} stored.
+     * <p>
+     * Note that there's a window of opportunity where the queue might exceed the {@code maxQueues} value to accompany
+     * concurrent usage of this dead-letter queue.
+     *
+     * @return The maximum number of distinct queues this dead-letter queue can hold.
+     */
+    long maxQueues();
+
+    /**
+     * The maximum number of letters a single queue can contain. A single queue is referenced based on it's
+     * {@link QueueIdentifier}.
+     * <p>
+     * Note that there's a window of opportunity where the queue might exceed the {@code maxQueueSize} value to
+     * accompany concurrent usage.
+     *
+     * @return The maximum number of letters a single queue can contain.
+     */
+    long maxQueueSize();
+
+    /**
+     * Take the oldest {@link DeadLetter} from this dead-letter queue for the given {@code group} that is ready to be
+     * released. Letter sequences can be made available earlier through {@link #release(Predicate)} when necessary.
+     * <p>
+     * Upon taking, the returned {@link DeadLetter dead-letter} is kept in the queue with an updated
+     * {@link DeadLetter#expiresAt()} and {@link DeadLetter#numberOfRetries()}. Doing so guards the queue against
+     * concurrent take operations accidentally retrieving (and thus handling) the same letter.
+     * <p>
+     * Will return an {@link Optional#empty()} if there are no letters ready to be released (i.e., they have not
+     * {@link DeadLetter#expiresAt() expired} yet) or present for the given {@code group}.
+     *
+     * @param group The group descriptor of a {@link QueueIdentifier} to take a letter for.
+     * @return The oldest {@link DeadLetter} belonging to the given {@code group} from this dead-letter queue.
+     * @see #release(Predicate)
+     */
+    Optional<DeadLetter<T>> take(@Nonnull String group);
+
+    /**
+     * Release all {@link DeadLetter dead-letters} within this queue that match the given {@code queueFilter}.
+     * <p>
+     * This makes the matching letters ready to be {@link #take(String) taken}. Furthermore, it signals any matching
+     * (based on the {@code group} name) callbacks registered through {@link #onAvailable(String, Runnable)}.
+     *
+     * @param queueFilter A lambda selecting the letters within this queue to be released.
+     */
+    void release(@Nonnull Predicate<QueueIdentifier> queueFilter);
```

Comment:
We should make clear that concurrently `take`ing and `release`ing can have consequences. The `take` operations are safe to execute concurrently (as written in the Javadoc), but they are not safe to concurrently execute with `release`, as this may cause two `take` operations to take the same entry.

Reply:
The `take` and `release` operations have been replaced for the `process` operation. As the sole means to retrieve, taking into account concurrent access, the predicament should be eliminated.